src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -97,6 +97,13 @@ internal class CodeActionEndpoint : IVSCodeActionEndpoint
         {
             _supportsCodeActionResolve = clientCapabilities.TextDocument?.CodeAction?.ResolveSupport != null;
 
+            // VSCode registers built-in features by default:
+            // https://github.com/microsoft/vscode-languageserver-node/blob/ed6a6d7da0ad64ebea0b55e4b2f339a1ec7f511f/client/src/common/client.ts#L1615
+            if (!clientCapabilities.SupportsVisualStudioExtensions)
```

Comment:
I don't like this way of "detecting" VS Code. If VS code ever supports VS extensions, would we then want to register it? What about if VS Mac doesn't support VS extensions, but doesn't have built-in features that register by default?

We should use `LanguageServerFeatureOptions` for it, because already have [Windows](https://github.com/dotnet/razor-tooling/blob/24a357bb0a4e18064ba96faa824ec3f7c612d6dc/src/Razor/src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/VisualStudioMacLanguageServerFeatureOptions.cs), [Mac](https://github.com/dotnet/razor-tooling/blob/24a357bb0a4e18064ba96faa824ec3f7c612d6dc/src/Razor/src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/VisualStudioMacLanguageServerFeatureOptions.cs) and [Default](https://github.com/dotnet/razor-tooling/blob/24a357bb0a4e18064ba96faa824ec3f7c612d6dc/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer.Common/DefaultLanguageServerFeatureOptions.cs) versions. We actually already have a VS Code one kinda, [here](https://github.com/dotnet/razor-tooling/blob/24a357bb0a4e18064ba96faa824ec3f7c612d6dc/src/Razor/src/Microsoft.AspNetCore.Razor.OmniSharpPlugin.StrongNamed/OmniSharpLanguageServerFeatureOptions.cs#L11) so I guess that should be made a little more official, and not just delegate to the default one. Though having said that, I'm not entirely sure if the default one is ever actually used then, but cleaning that up could be a follow up tech-debt issue.

Reply:
Yeah I think I agree about "detecting" VSCode. I also feel like we shouldn't be artificially restricting the capabilities our LS reports for the benefit of particular clients. If we imagine having 3-5+ clients that use our LS the dangers of conditionalizing all this become quite messy. Where possible I feel we should push all introspecting about the client out of the LS and instead only consider its reported capabilities. In this particular case just because the VSCode client registers this feature already I don't see any reason we shouldn't be reporting it available anyway.

Reply:
> I don't see any reason we shouldn't be reporting it available anyway.

*And then expecting the VSCode client to know not to re-register them.

Reply:
>  In this particular case just because the VSCode client registers this feature already I don't see any reason we shouldn't be reporting it available anyway.

Are you saying we should leave our capabilities code alone and filter these out on the VS Code side? or just report a bug in VS Code? If so, i'm in support.

Reply:
Basically both of those suggestions then revert once they fix (or advise on a better way).

Reply:
I took David's suggestion and created a new `LanguageServerFeatureOptions` property to use here. Unfortunately, due to the way `LanguageServerFeatureOptions` are passed into the Razor language server, we can't actually recognize which client we're dealing with when initializing features except in the case of VS Windows/Mac. I added a workaround for now, but long term we probably just want VS Code to fix the issue in case any other clients are added.

> Are you saying we should leave our capabilities code alone and filter these out on the VS Code side? 

I don't think we can do that from our end. The Razor language server from the Razor VSCode POV is a "black box" in the sense that we don't really have control over what's going on within the language server internally (besides sending requests/receiving responses).

> or just report a bug in VS Code?

Created https://github.com/microsoft/vscode-languageserver-node/issues/1075